File: src/android/app/src/main/java/org/citra/citra_emu/model/GameInfo.java
package org.citra.citra_emu.model;

import androidx.annotation.Keep;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.IOException;

public class GameInfo {
    @Keep
    private final long mPointer;

    @Keep
    public GameInfo(String path) throws IOException {
        mPointer = initialize(path);
        if (mPointer == 0L) {
            throw new IOException();
        }
    }

    private static native long initialize(String path);

    @Override
    protected native void finalize();

    @NonNull
    public native String getTitle();

    @NonNull
    public native String getRegions();

    @NonNull
    public native String getCompany();

    @Nullable
    public native int[] getIcon();
}

File: src/android/app/src/main/java/org/citra/citra_emu/model/GameInfo.kt
// Copyright 2023 Citra Emulator Project
// Licensed under GPLv2 or any later version
// Refer to the license.txt file included.

package org.citra.citra_emu.model

import androidx.annotation.Keep
import java.io.IOException

class GameInfo(path: String) {
    @Keep
    private val pointer: Long

    init {
        pointer = initialize(path)
        if (pointer == 0L) {
            throw IOException()
        }
    }

    protected external fun finalize()

    external fun getTitle(): String

    external fun getRegions(): String

    external fun getCompany(): String

    external fun getIcon(): IntArray?

    external fun getIsVisibleSystemTitle(): Boolean

    companion object {
        @JvmStatic
        private external fun initialize(path: String): Long
    }
}
